src/gravel/controllers/deployment/mgr.py

```diff
@@ -0,0 +1,635 @@
+        except FileExistsError:
+            loc = confdir.joinpath("state.json")
+            raise InitError(f"On-disk state is not a file at '{loc}'")
+
+        self._init_state = state.init
+        self._deployment_state = state.deployment
+        self._inited = True
+
+    def postinit(self, gstate: GlobalState, nodemgr: NodeMgr) -> None:
+        """Called after all state has been inited, prepares for deployment."""
+        assert gstate is not None
+        assert gstate.ready
+        assert nodemgr is not None
+        assert self._running
+        assert self._task_install is None
+        assert self._init_state >= InitStateEnum.INSTALLED
+        logger.debug("Post-init Deployment Manager.")
+        self._gstate = gstate
+        self._nodemgr = nodemgr
+        self._postinited = True
+
+        if self._deployment_state == DeploymentStateEnum.DEPLOYED:
+            self._mark_deployed()
+
+    def _mark_deployed(self) -> None:
+        logger.debug("Marking Node Manager Deployed.")
+        if self._nodemgr is None:
+            logger.error("Error: We don't have a Node Manager yet!")
+            return
+        self._nodemgr.mark_deployed()
+
+    def get_status(self) -> DeploymentStatusModel:
+        """Obtain node deployment status."""
+        progress: Optional[ProgressModel] = None
+
+        if self._creator is not None:
+            create = self._creator.progress
+            progress = ProgressModel(value=create.progress, msg=create.msg)
+        elif self._join_requester is not None:
+            joiner = self._join_requester.progress
+            progress = ProgressModel(value=joiner.progress, msg=joiner.msg)
+
+        return DeploymentStatusModel(
+            state=DeploymentStateModel(
+                init=self._init_state,
+                deployment=self._deployment_state,
+            ),
+            progress=progress,
+            error=self._error,
+        )
+
+    async def get_requirements(self) -> RequirementsModel:
+        """Obtain node requirements."""
+        return await localhost_qualified()
+
+    async def get_devices(self) -> List[DiskDevice]:
+        """Obtain storage devices."""
+        try:
+            return await get_storage_devices()
+        except Exception:
+            return []
+
+    async def install(self, device: str) -> None:
+        """Install Aquarium onto a specified device."""
+
+        logger.debug(f"Attempt installing on {device}.")
+        if self.installed:
+            raise NodeInstalledError("Node is already installed.")
+        if self._deployment_state >= DeploymentStateEnum.INSTALLING:
+            logger.info("Node is currently installing; no op.")
+            return
+
+        assert self._task_main is not None
+        assert not self._task_main.done()
+        assert self._task_install is None
+
+        # reset error, if any
+        self._error = DeploymentErrorModel()
+        self._task_install = asyncio.create_task(self._install_task(device))
```

Comment:
Exceptions thrown in the task are not catched, thus executing `/api/deploy/install` with an device that does not exist will always return `{success:true}`.

Reply:
That's correct. We're not throwing any exceptions through that function, just setting error messages in case there's an error. The only way to properly deal with that is by calling the `/api/deploy/install` endpoint and then check the `/api/deploy/status` endpoint for errors.